File: src/main/java/ru/serdtsev/homemoney/balancesheet/BalanceSheet.java
package ru.serdtsev.homemoney.balancesheet;

import com.fasterxml.jackson.annotation.JsonIgnore;
import ru.serdtsev.homemoney.account.model.*;

import javax.annotation.Nullable;
import javax.persistence.*;
import java.io.Serializable;
import java.sql.Date;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Entity
@Table(name = "balance_sheets")
public class BalanceSheet implements Serializable {
  @Id
  private UUID id;

  private Timestamp created;

  @Column(name = "currency_code")
  private String currencyCode;

  @OneToOne
  @JoinColumn(name = "svc_rsv_id", insertable = false)
  private ServiceAccount svcRsv;

  @OneToOne
  @JoinColumn(name = "uncat_costs_id", insertable = false)
  private Category uncatCosts;

  @OneToOne
  @JoinColumn(name = "uncat_income_id", insertable = false)
  private Category uncatIncome;

  @OneToMany
  @JoinColumn(name = "balance_sheet_id")
  private List<Account> accounts;

  @SuppressWarnings({"unused", "WeakerAccess"})
  protected BalanceSheet() { }

  private BalanceSheet(UUID id, Timestamp created, String currencyCode) {
    this(id, created, currencyCode, null, null, null);
  }

  private BalanceSheet(UUID id, Timestamp created, String currencyCode, @Nullable ServiceAccount svcRsv,
      @Nullable Category uncatCosts, @Nullable Category uncatIncome) {
    this.id = id;
    this.created = created;
    this.currencyCode = currencyCode;
    this.svcRsv = svcRsv;
    this.uncatCosts = uncatCosts;
    this.uncatIncome = uncatIncome;
  }

  public UUID getId() {
    return id;
  }

  public static BalanceSheet newInstance() {
    return new BalanceSheet(UUID.randomUUID(), Timestamp.valueOf(LocalDateTime.now()), "RUB")
        .init();
  }

  public BalanceSheet init() {
    Date now = Date.valueOf(LocalDate.now());
    svcRsv = new ServiceAccount(this, "Service reserve", now, false);
    uncatCosts = new Category(this, AccountType.expense, "<Без категории>", now, false, null);
    uncatIncome = new Category(this, AccountType.income, "<Без категории>", now, false, null);
    return this;
  }


  public String getCurrencyCode() {
    return currencyCode;
  }

  @JsonIgnore
  public ServiceAccount getSvcRsv() {
    return svcRsv;
  }

  @JsonIgnore
  public Category getUncatCosts() {
    return uncatCosts;
  }

  @JsonIgnore
  public Category getUncatIncome() {
    return uncatIncome;
  }

  @JsonIgnore
  public List<Account> getAccounts() {
    return this.accounts;
  }

  @JsonIgnore
  public List<Balance> getBalances() {
    return this.accounts.stream()
        .filter(account -> account instanceof Balance)
        .map(account -> (Balance) account)
        .collect(Collectors.toList());
  }

  @Override
  public String toString() {
    return "BalanceSheet{" +
        "id=" + id +
        ", created=" + created +
        ", currencyCode='" + currencyCode + '\'' +
        '}';
  }
}

File: src/main/java/ru/serdtsev/homemoney/balancesheet/BalanceSheet.kt
package ru.serdtsev.homemoney.balancesheet

import com.fasterxml.jackson.annotation.JsonIgnore
import ru.serdtsev.homemoney.account.model.*
import java.io.Serializable
import java.sql.Date
import java.time.Instant
import java.time.LocalDate
import java.util.*
import javax.persistence.*

@Entity
@Table(name = "balance_sheets")
data class BalanceSheet(
        @Id
        var id: UUID? = null,

        var created: Instant? = null,

        @Column(name = "currency_code")
        var currencyCode: String? = null,

        @OneToOne @JoinColumn(name = "svc_rsv_id", insertable = false) @get:JsonIgnore
        var svcRsv: ServiceAccount? = null,

        @OneToOne @JoinColumn(name = "uncat_costs_id", insertable = false) @get:JsonIgnore
        var uncatCosts: Category? = null,

        @OneToOne @JoinColumn(name = "uncat_income_id", insertable = false) @get:JsonIgnore
        var uncatIncome: Category? = null,

        @OneToMany
        @JoinColumn(name = "balance_sheet_id")
        @get:JsonIgnore
        var accounts: MutableList<Account>? = null
) : Serializable {
    val balances: List<Balance>?
        @JsonIgnore get() = this.accounts
                ?.filter { account -> account is Balance }
                ?.map { account -> account as Balance }
                .orEmpty()

    fun init(): BalanceSheet {
        val now = Date.valueOf(LocalDate.now())
        svcRsv = ServiceAccount(this, "Service reserve", now, false)
        uncatCosts = Category(this, AccountType.expense, "<Без категории>", now, false, null)
        uncatIncome = Category(this, AccountType.income, "<Без категории>", now, false, null)
        return this
    }

    override fun toString() = "BalanceSheet{id=$id, created=$created, currencyCode=$currencyCode}"

    companion object {
        fun newInstance(): BalanceSheet {
            return BalanceSheet(UUID.randomUUID(), Instant.now(), "RUB").init()
        }
    }
}
